x-pack/plugin/search-business-rules/src/main/java/org/elasticsearch/xpack/searchbusinessrules/RuleQueryBuilder.java

```diff
@@ -0,0 +1,361 @@
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.searchbusinessrules;/*
+                                                    * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+                                                    * or more contributor license agreements. Licensed under the Elastic License
+                                                    * 2.0 and the Server Side Public License, v 1; you may not use this file except
+                                                    * in compliance with, at your election, the Elastic License 2.0 or the Server
+                                                    * Side Public License, v 1.
+                                                    */
+
+import org.apache.lucene.search.Query;
+import org.apache.lucene.util.SetOnce;
+import org.elasticsearch.TransportVersion;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.common.ParsingException;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.index.query.AbstractQueryBuilder;
+import org.elasticsearch.index.query.BoolQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryBuilders;
+import org.elasticsearch.index.query.QueryRewriteContext;
+import org.elasticsearch.index.query.SearchExecutionContext;
+import org.elasticsearch.logging.LogManager;
+import org.elasticsearch.logging.Logger;
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Supplier;
+
+import static org.elasticsearch.xcontent.ConstructingObjectParser.constructorArg;
+
+/**
+ * A query that will determine based on query context and configured query rules,
+ * whether a query should be modified based on actions specified in matching rules.
+ *
+ * This iteration will determine if a query should have pinned documents and if so,
+ * modify the query accordingly to pin those documents.
+ */
+public class RuleQueryBuilder extends AbstractQueryBuilder<RuleQueryBuilder> {
```

Comment:
I wonder if we should restrict this type of query to only be allowed when expensive queries are enabled, similar to what we do with percolator. WDYT?

Reply:
We probably want to set up a feature flag for this, so we may decide about it later